riscv/trace_encoder_n.cc

```diff
@@ -0,0 +1,211 @@
+#include "trace_encoder_n.h"
+
+trace_encoder_n::trace_encoder_n() {
+  this->trace_sink= fopen("trace_n.bin", "wb");
+  this->debug_reference = fopen("trace_n_ref_debug.log", "wb");
+  this->active = true;
+  this->enabled = false;
+  this->src = 0;
+  this->state = TRACE_ENCODER_N_IDLE;
+  this->icnt = 0;
+  this->packet_0 = (hart_to_encoder_ingress_t*) malloc(sizeof(hart_to_encoder_ingress_t)); // create empty packet
+  this->packet_1 = (hart_to_encoder_ingress_t*) malloc(sizeof(hart_to_encoder_ingress_t)); // create empty packet
```

Comment:
These don't need to be pointers again, these can just be member objects. I suspect the overhead of your repeated `malloc`/`free` in the inner loop will offset any benefit from the pass-by-reference, if performance is what you are concerned about.